### PR TITLE
feat!: added computed members support

### PR DESCRIPTION
**Pull request by SpecialistSun**
- Added computed members support
- Updated docs
- Updated test units
- Updated changelog

### CODE REVIEW

File: variables.tf [hunk 1]
```diff
@@ -25,11 +25,22 @@ variable "members" {
...
   }
 }
 
+variable "computed_members_map" {
+  type        = map(string)
+  description = "(Optional) A map of members to replace in 'var.members' or in members of 'policy_bindings' to handle terraform computed values."
+  default     = {}
+
+  validation {
+    condition     = alltrue([for k, v in var.computed_members_map : can(regex("^(allUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|principalSet|principal):)", v))])
```
Comment (waxb) [inline variables.tf:39]: please update validation condition accordingly to the [documentation](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloudfunctions_function_iam)

Comment (waxb) [inline variables.tf:39]: please update validation condition accordingly to the [documentation](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloudfunctions_function_iam)

**Review by waxb — CHANGES_REQUESTED** [review]

File: terramate.tm.hcl [hunk 2]
```diff
@@ -0,0 +1,24 @@
+stack {
+  name        = "terraform-module-template"
```
Comment (switchdk) [inline terramate.tm.hcl:2]: just noticed this. Remember to update the stack `name` and the `description` just below after copying from the template :smile_cat:

File: test/unit-minimal/main.tf [hunk 3]
```diff
@@ -9,7 +9,5 @@ module "test" {
 
   role = "roles/viewer"
```
Comment (waxb) [inline test/unit-minimal/main.tf:10]: you can remove this as it is optional

File: test/unit-complete/main.tf [hunk 4]
```diff
@@ -1,11 +1,86 @@
...
-#   source = "../.."
+module "test-sa" {
+  source = "github.com/mineiros-io/terraform-google-service-account?ref=v0.1.0"
 
-#   module_enabled = false
+  account_id = "service-account-id-${local.random_suffix}"
+}
 
-#   # add all required arguments
+module "test" {
```
Comment (waxb) [inline test/unit-complete/main.tf:7]: please add all arguments to test complete

**Review by waxb — CHANGES_REQUESTED** [review]
couple of small things remaining:
- align `outputs.tf` with the example repo
- change pr title to state computed members change
- have three commits: computed members, test update, changelog
- changelog :)

File: test/unit-minimal/main.tf (hunk 3 above)
Comment (waxb) [reply test/unit-minimal/main.tf:10]: will discuss with @mariux shortly :)

File: test/unit-complete/main.tf [hunk 5]
```diff
@@ -1,11 +1,117 @@
-# module "test" {
-#   source = "../.."
+module "test-sa" {
+  source = "github.com/mineiros-io/terraform-google-service-account?ref=v0.1.0"
```
Comment (switchdk) [inline test/unit-complete/main.tf:2]:
```suggestion
  source = "github.com/mineiros-io/terraform-google-service-account?ref=v0.1.1"
```

File: README.md [hunk 6]
```diff
@@ -106,9 +86,17 @@ See [variables.tf] and [examples/] for details and use-cases.
   - `projectOwner:projectid`: Owners of the given project. For example, `projectOwner:my-example-project`
   - `projectEditor:projectid`: Editors of the given project. For example, `projectEditor:my-example-project`
   - `projectViewer:projectid`: Viewers of the given project. For example, `projectViewer:my-example-project`
+  - `computed:{identifier}`: An existing key from var.computed_members_map.
 
   Default is `[]`.
 
+- [**`computed_members_map`**](#var-computed_members_map): *(Optional `map(string)`)*<a name="var-computed_members_map"></a>
+
+  A map of identifiers to identities to be replaced in 'var.members' or in members of `policy_bindings` to handle terraform computed values.
```
Comment (switchdk) [inline README.md:95]:
```suggestion
  A map of identifiers to identities to be replaced in `var.members` or in members of `policy_bindings` to handle terraform computed values.
```

File: README.tfdoc.hcl [hunk 7]
```diff
@@ -140,9 +116,29 @@ section {
...
             - `projectViewer:projectid`: Viewers of the given project. For example, `projectViewer:my-example-project`
+            - `computed:{identifier}`: An existing key from var.computed_members_map.
           END
         }
 
+        variable "computed_members_map" {
+          type        = map(string)
+          default     = {}
+          description = <<-END
+             A map of identifiers to identities to be replaced in 'var.members' or in members of `policy_bindings` to handle terraform computed values.
```
Comment (switchdk) [inline README.tfdoc.hcl:127]:
```suggestion
             A map of identifiers to identities to be replaced in `var.members` or in members of `policy_bindings` to handle terraform computed values.
```

File: variables.tf [hunk 8]
```diff
@@ -25,11 +25,22 @@ variable "members" {
...
   validation {
-    condition     = alltrue([for m in var.members : can(regex("^(allUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|projectOwner|projectEditor|projectViewer):)", m))])
+    condition     = alltrue([for m in var.members : can(regex("^(allUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|projectOwner|projectEditor|projectViewer|computed):)", m))])
     error_message = "The value must be a non-empty list of strings where each entry is a valid principal type identified with `allUsers`, `allAuthenticatedUsers` or prefixed with `user:`, `serviceAccount:`, `group:`, `domain:`, `projectOwner:`, `projectEditor:` or `projectViewer:`."
   }
 }
 
+variable "computed_members_map" {
+  type        = map(string)
+  description = "(Optional) A map of members to replace in 'var.members' or in members of 'policy_bindings' to handle terraform computed values."
```
Comment (switchdk) [inline variables.tf:35]:
```suggestion
  description = "(Optional) A map of members to replace in `var.members` or in members of `policy_bindings` to handle terraform computed values."
```

File: variables.tf [hunk 9]
```diff
@@ -25,11 +25,22 @@ variable "members" {
...
+  default     = {}
+
+  validation {
+    condition     = alltrue([for k, v in var.computed_members_map : can(regex("^(allUsers|allAuthenticatedUsers|(llUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|projectOwner|projectEditor|projectViewer):)", v))])
+    error_message = "The value must be a non-empty string being a valid principal type identified with ``allUsers`, `allAuthenticatedUsers` or prefixed with `user:`, `serviceAccount:`, `group:`, `domain:`, `projectOwner:`, `projectEditor:` or `projectViewer:`."
+  }
+}
+
 variable "role" {
   description = "(Optional) The role that should be applied. Only one 'iam_binding' can be used per role. Note that custom roles must be of the format '[projects|organizations]/{parent-name}/roles/{role-name}'."
```
Comment (switchdk) [inline variables.tf:45]:
```suggestion
  description = "(Optional) The role that should be applied. Only one `iam_binding` can be used per role. Note that custom roles must be of the format `[projects|organizations]/{parent-name}/roles/{role-name}`."
```

File: test/unit-complete/main.tf [hunk 10]
```diff
@@ -1,11 +1,117 @@
...
+    "serviceAccount:service-account-name@${local.project_id}.iam.gserviceaccount.com",
+    "group:group@${local.org_domain}",
+    "domain:${local.org_domain}",
+    "projectOwner:${local.project_id}",
+    "projectEditor:${local.project_id}",
+    "projectViewer:${local.project_id}",
+  ]
+
+  # add most/all other optional arguments
+
```
Comment (switchdk) [inline test/unit-complete/main.tf:31]: add `computed_members_map`

File: variables.tf [hunk 11]
```diff
@@ -25,11 +25,22 @@ variable "members" {
   default     = []
 
   validation {
-    condition     = alltrue([for m in var.members : can(regex("^(allUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|projectOwner|projectEditor|projectViewer):)", m))])
+    condition     = alltrue([for m in var.members : can(regex("^(allUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|projectOwner|projectEditor|projectViewer|computed):)", m))])
     error_message = "The value must be a non-empty list of strings where each entry is a valid principal type identified with `allUsers`, `allAuthenticatedUsers` or prefixed with `user:`, `serviceAccount:`, `group:`, `domain:`, `projectOwner:`, `projectEditor:` or `projectViewer:`."
```
Comment (switchdk) [inline variables.tf:29]:
```suggestion
    error_message = "The value must be a non-empty list of strings where each entry is a valid principal type identified with `allUsers`, `allAuthenticatedUsers` or prefixed with `user:`, `serviceAccount:`, `group:`, `domain:`, `projectOwner:`, `projectEditor:`, `projectViewer:` or  `computed:`."
```

**Review by waxb — CHANGES_REQUESTED** [review]
please add [terramate manifest](https://github.com/mineiros-io/terraform-module-template/tree/main/.terramate) as well

**Review by waxb — CHANGES_REQUESTED** [review]
please add [terramate manifest](https://github.com/mineiros-io/terraform-module-template/tree/main/.terramate)

**Comment by waxb** [comment]
this is a breaking change as support for provider version 3 has been dropped. please put this into the changelog and update the title

File: main.tf [hunk 12]
```diff
@@ -16,7 +22,7 @@ resource "google_cloudfunctions_function_iam_member" "member" {
   cloud_function = var.cloud_function
   region         = var.region
   role           = var.role
-  member         = each.value
+  member         = try(var.computed_members_map[regex("^computed:(.*)", each.value)[0]], each.value)
```
Comment (waxb) [inline main.tf:25]: please update the `google_cloudfunctions_function_iam_policy` for computed members as well

**Review by waxb — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 13]
```diff
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- Add computed members support
+
+### Removed
+
+- BREAKING: Remove support for Terraform Google Provider versions before v4.0
```
Comment (waxb) [inline CHANGELOG.md:16]: also terraform version >1 dropped

**Review by waxb — CHANGES_REQUESTED** [review]
changelog update is not a breaking commit

**Review by waxb — APPROVED** [review]
lgtm

**Review by mariux — DISMISSED** [review]
lgtm, good job, thanks.

File: variables.tf [hunk 14]
```diff
@@ -25,13 +25,24 @@ variable "members" {
...
+  }
+}
+
+variable "computed_members_map" {
+  type        = map(string)
+  description = "(Optional) A map of members to replace in `var.members` or in members of `policy_bindings` to handle terraform computed values."
+  default     = {}
+
+  validation {
+    condition     = alltrue([for k, v in var.computed_members_map : can(regex("^(allUsers|allAuthenticatedUsers|(llUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|projectOwner|projectEditor|projectViewer):)", v))])
```
Comment (switchdk) [inline variables.tf:39]:
```suggestion
    condition     = alltrue([for k, v in var.computed_members_map : can(regex("^(allUsers|allAuthenticatedUsers|(user|serviceAccount|group|domain|projectOwner|projectEditor|projectViewer):)", v))])
```

**Review by switchdk — CHANGES_REQUESTED** [review]
Sorry, I missed these changes during my first review. I noticed that a few changes are missing in a few places based on other updated modules. If my notes don't make sense, please just let me know. I wasn't able to comment directly in on the files in Github so I had to make notes like the below instead.

1. `main.tf`, line 20 (`count =`). Comparing to `terraform-google-storage-bucket-iam`, [line 28](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/main.tf#L28) 
2. `main.tf`, line 32 (`count =`). Comparing to `terraform-google-storage-bucket-iam`, [line 48](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/main.tf#L48) 
3. `main.tf`, line 36 (`policy_data`). Comparing to `terraform-google-storage-bucket-iam`, [line 51](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/main.tf#L51) 
4. `main.tf`, line 43 (`count =`), Comparing to `terraform-google-bucket-storage-iam`, [line 57](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/main.tf#L57)
5. I think `terramate generate` has not been executed since the unit test folders are missing the `_generated_google.tf` files. If `terramate generate` has been run but the files are not generated then `test/terramate_google.tm.hcl` is probably missing
6. Please update the `test/unit_*_test.go` files. They should contain the three env vars for project, domain and billing account.
7. In `variables.tf`, line 24, the `description` contains single quotes `'` instead of backticks `\``. Please replace with backticks (just like you did in `variable role`
8. In `variables.tf, line 75 and 81. you can remove the `Default is...` in the `description` since it is included in the Terradoc template.

File: main.tf [hunk 15]
```diff
@@ -41,7 +47,7 @@ data "google_iam_policy" "policy" {
```
Comment (mariux) [inline main.tf:47]: please use the new local create_policy for count condition here..

File: main.tf [hunk 16]
```diff
@@ -16,7 +22,7 @@ resource "google_cloudfunctions_function_iam_member" "member" {
   cloud_function = var.cloud_function
```
Comment (mariux) [inline main.tf:22]: please use the new local create_member in for_each condition in this resource

**Review by mariux — CHANGES_REQUESTED** [review]

**Comment by mariux** [comment]
> Sorry, I missed these changes during my first review. I noticed that a few changes are missing in a few places based on other updated modules. If my notes don't make sense, please just let me know. I wasn't able to comment directly in on the files in Github so I had to make notes like the below instead.
> 
>     1. `main.tf`, line 20 (`count =`). Comparing to `terraform-google-storage-bucket-iam`, [line 28](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/main.tf#L28)
> 
>     2. `main.tf`, line 32 (`count =`). Comparing to `terraform-google-storage-bucket-iam`, [line 48](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/main.tf#L48)
> 
>     3. `main.tf`, line 36 (`policy_data`). Comparing to `terraform-google-storage-bucket-iam`, [line 51](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/main.tf#L51)
> 
>     4. `main.tf`, line 43 (`count =`), Comparing to `terraform-google-bucket-storage-iam`, [line 57](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/main.tf#L57)

i agree added some notes inline.. please fix those 1-4 topics.

>     7. In `variables.tf`, line 24, the `description` contains single quotes `'` instead of backticks. Please replace with backticks (just like you did in `variable role`

that;s a ice to have, not a blocker though ;) nice catch.

>     8. In `variables.tf`, line 75 and 81. you can remove the `Default is...` in the `description` since it is included in the Terradoc template.

yes, defaults can be omitted... nice catch.. not a blocker...

**Review by switchdk — APPROVED** [review]
Great stuff @SpecialistSun :heart_eyes_cat:

**Review by waxb — APPROVED** [review]
lgtm!

**Review by mariux — APPROVED** [review]
lgtm

**Pull request merged**